Cargo.toml

```
@@ -17,10 +17,21 @@ ring = "0.16.20"
 time = "0.2.26"
 percent-encoding = "2.1.0"
 thiserror = "1.0.24"
+http="0.2"
+url="2"
+
+[features]
+client-reqwest=["reqwest"]
+reqwest-blocking=["client-reqwest"]
+default=["client-reqwest", "reqwest-blocking"]
 
 [dependencies.reqwest]
+optional=true
 version = "0.11.3"
 features = ["blocking"]
```

Can you enable the `blocking` feature for `reqwest` only if `reqwest-blocking` is enabled for `oauth-client`?
For example, how about the following (this fix may not be enough to make the source code compile)

```suggestion
[features]
client-reqwest=["reqwest"]
reqwest-blocking=["reqwest/blocking"]
default=["client-reqwest", "reqwest-blocking"]

[dependencies.reqwest]
optional=true
version = "0.11.3"
```

---

> Can you enable the blocking feature for reqwest only if reqwest-blocking is enabled for oauth-client?

Yes, the idea being you might want to use reqwest but with async.

> For example, how about the following 

I was not aware you can do this `"reqwest/blocking"`.

---

I read the "Dependency features" and "feature unification" chapters of https://doc.rust-lang.org/cargo/reference/features.html#dependency-features and as I understand the issue will only happen if e.g. you're using oauth-client-rs with features: `["client-reqwest"]` and also you're using a dependency that itself is using oauth-client-rs with a different set of features that does not include the `client-reqwest` feature, because Cargo will enable all features the `cfg(not)` branch will never get called.

I think this is unlikely, so the `compile_error` definitely needs to be added. But for ultimate correctness I think the crate should be split into three workspaces/crates with a `core` one, one for `reqwest` and one for a `custom` client. Not sure if it's worth the trouble and confusion though.

---

In https://github.com/gifnksm/oauth-client-rs/pull/44#discussion_r685036991 ,   `#[cfg(not(...))]` is removed, so you can enable `reqwest/blocking` without compile errors [like this commit].

[like this commit]: https://github.com/gifnksm/oauth-client-rs/commit/581b8a6aaf08e040e78d158ba07f5a84c5a56506

If we do not adopt the above fix, I think adding `custom-client` feature is better.

In general, it's not good to add an implementation when a feature is disabled. This is because if other dependent crates enable the feature, no implementations will be added.

---

> If we do not adopt the above fix,

I see no reason why not. Looks good to me.